smolyay/grid.py

```diff
@@ -0,0 +1,458 @@
+        self._basis = self._nested_basis_set.basis_functions
+        self._levels = self._nested_basis_set.levels
+
+    @property
+    def nested_basis_set(self):
+        """:class:`NestedBasisSet`: Nested basis set (data structure)."""
+        return self._nested_basis_set
+
+    @nested_basis_set.setter
+    def nested_basis_set(self, nested_basis_set):
+        self._nested_basis_set = nested_basis_set
+
+    def indices(self, dimension, expand_indicy=[], drop_indicy=[]):
+        """Generate Smolyak indices depending on the dimensionality.
+
+        Parameters
+        ----------
+        dimension: int
+            Number of independent variables.
+        expand_indicy: iterator
+            Indicy that is needed to be added.
+        drop_indicy: iterator
+            Indicy that is needed to be dropped.
+
+        Returns
+        -------
+        list
+            Smolyak indices.
+
+        Raises
+        ------
+        IndexError
+            Indicy that is needed to be added or dropped must be in
+            length of the dimension.
+        """
+        self._expand_indicy = expand_indicy
+        self._drop_indicy = drop_indicy
+        self._update(dimension)
+
+        if ((self._drop_indicy != [] and len(self._drop_indicy) != dimension)
+            or (self._expand_indicy != [] and
+                len(self._expand_indicy) != dimension)):
+            raise IndexError("Indicy must be in length of the dimension.")
+
+        if (self._drop_indicy != [] and
+           self._drop_indicy not in self._indices):
+            raise ValueError("Indicy doesn't exist.")
+
+        if self._expand_indicy == [] and self._drop_indicy == []:
+            return self._indices
+        else:
+            if self._expand_indicy == []:
+                self._indices.remove(drop_indicy)
+            elif self._drop_indicy == []:
+                self._indices.append(expand_indicy)
+            else:
+                self._indices.append(expand_indicy)
+                self._indices.remove(drop_indicy)
+            return self._indices
+
+    def __call__(self, dimension):
+        """Make grid points and their corresponding basis functions.
+
+        Depending on the dimensionality, a set of grid points are generated
+        (integer grid points). Grid points of the basis function, and
+        their corresponding basis function are then generated by replacing
+        the indexes with their points and functions.
+
+        Parameters
+        ----------
+        dimension: int
+            Number of independent variables.
+
+        Returns
+        -------
+        :class:`SmolyakGrid`
+            Grid points (and their indexes),their corresponding functions and
+            indices.
+        """
+        self._dimension = dimension
+        self._update(self._dimension)
+
+        return SmolyakGrid(self._grid_points_indexes, self._grid_points,
+                           self._grid_points_basis, self._indices)
```

Comment:
Similarly here, you should not be saving these into attributes. Get rid of `_update`, and put those implementation parts here. These should be local variables that you use to return this grid.